[김은혜] Step3 질문 삭제하기 리팩터링 PR 요청

eunveloper opened the pull request:
벌써 JPA 마지막 단계네요 :)

과제 구현하기 전에 첨부되어있는 강의를 들었는데, 저도 비지니스 로직을 서비스 레이어에 구현하고
나중에 리팩터링하거나 유지보수 했을때 힘들었던 경험이 있었었어 많이 느꼈어요!

이번과제에서 궁금증이 있는부분 하나 질문 남겨요
Entity를 생성할때 생성자로 넘어오는 문자열 데이터들도 포장하여 사용하는것이 좋을까요?

예를 든다면
`public Question(Long id, String title, String contents) {
        this.id = id;
        this.title = title;
        this.contents = contents;
}`
이런 코드에서 Title이라는 객체와 Content라는 객체를 사용하는게 좋은 방향인지
아니면 이런 순수한 데이터 세팅용의 부분에는 그냥 문자열을 주고받는게 좋은 방향인지 궁금합니다!

저는 개인적으로 비지니스 로직내에서 연산이 이뤄지는게 아니라면 (즉, 객체에 연산에 관련한 메세지 전달이 필요없다면)
굳이 포장하지 않아도 생각했는데, 또 다른 생각이 있으신지 궁금해서요 :)

아무튼 JPA 3단계동안 많이 배워갑니다! 감사합니다~,~

eunveloper committed:
docs: README.md 작성

- 요구사항 작성
- 관련 sql문 작성

eunveloper committed:
feat: Entity 객체 sql 맞춰 구성

- 패키지 구조 세분화
- Entity 객체 JPA 어노테이션 적용
- 공통 시간값 관리하는 BaseTime 객체 생성

eunveloper committed:
feat: 테스트코드 작성

- AnswerTest 테스트 작성
- QuestionTest 테스트 작성
- UserTest 테스트 작성

eunveloper committed:
Merge remote-tracking branch 'origin/eunveloper' into eunveloper

# Conflicts:
#	src/main/java/qna/domain/BaseTime.java
#	src/main/java/qna/domain/DeleteHistory.java
#	src/main/java/qna/domain/Question.java
#	src/main/java/qna/domain/User.java
#	src/test/java/qna/domain/AnswerTest.java
#	src/test/java/qna/domain/QuestionTest.java
#	src/test/java/qna/domain/UserTest.java

eunveloper committed:
Merge remote-tracking branch 'origin/eunveloper' into eunveloper

eunveloper committed:
docs: README.md 작성

- 3단계 과제 요구사항 작성

eunveloper committed:
refactor: step2 피드백 반영

- 자동생성 기본메서드 컨벤션 맞추기
- 테스트 픽스처를 팩토리메서드 방식으로 변경
- 가독성 개선을 위한 가로정렬 형식맞추기
- 불필요 로컬변수 제거

eunveloper committed:
feat: 비지니스 로직을 Service Layer -> Domain 으로 이동

- README.md 해야하는 것 작성
- 불필요 테스트코드 제거
- QnaService -> Domain 으로 비지니스 로직 이관

eunveloper committed:
feat: 단위테스트가 가능한 기능들 테스트코드 작성

- 기능 요구사항중 중복되지 않는 요구사항의 단위테스트 작성

eunveloper committed:
refactor: 코드 리팩토링

- 프로그래밍 요구사항 맞추기
- 기타 추가 리팩토링이 필요한 부분 리팩토링하기

devyonghee commented on README.md:
```diff
@@ -107,4 +107,60 @@ alter table question
+### 힌트
+테스트하기 쉬운 부분과 테스트하기 어려운 부분을 분리해 테스트 가능한 부분만 단위 테스트한다.(O)
+
+## 해야할것
+1. 비지니스 로직을 Service Layer -> Domain 으로 이동 (O)
+2. 단위테스트가 가능한 기능들 테스트코드 작성 (O)
+  - 질문 데이터를 완전히 삭제하는 것이 아니라 데이터의 상태를 삭제 상태(deleted - boolean type)로 변경한다.(O)
+  - 로그인 사용자와 질문한 사람이 같은 경우 삭제할 수 있다. / 질문자와 답변자가 다른 경우 답변을 삭제할 수 없다.(O)
+  - 질문자와 답변 글의 모든 답변자 같은 경우 삭제가 가능하다.(O)
+  - 질문을 삭제할 때 답변 또한 삭제해야 하며, 답변의 삭제 또한 삭제 상태(deleted)를 변경한다.(O)
+  - 질문과 답변 삭제 이력에 대한 정보를 DeleteHistory를 활용해 남긴다.(O)
+3. 코드 리팩토링(O)
+  - 프로그래밍 요구사항 맞추기(O)
+  - 기타 추가 리팩토링이 필요한 부분 리팩토링하기(O)
```
요구사항 정리를 잘해주신 것 같아요!! 👍👍👍

devyonghee commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -36,17 +39,30 @@ public Answer(User writer, Question question, String contents) {
-        this.writer = writer;
-        this.question = question;
-        this.contents = contents;
+    public void isSameOwner(User loginUser) throws CannotDeleteException {
+        if (!isOwner(loginUser)) {
+            throw new CannotDeleteException(NOT_QUESTION_DELETE_WRITE_OTHER_USER);
+        }
```
도메인 내부의 에러 메세지를 외부 클래스의 상수로 선언할 정도로 범용성있게 재사용 하는지 의문이 드는 부분입니다. 🤔🤔
메세지로 인해 도메인에서 외부 클래스와의 결합도가 생긴 것 같은데요.
혹시 이 부분에 대해서 어떻게 생각하시나요?? 🤔

또한 어떠한 정보로 인해 예외가 발생된 것인지 자세한 내용을 추가해주는 것은 어떠신가요?? 🤔🤔
제가 참고한 자료 남겨드리도록 하겠습니다!
[예외의 상세 메시지에 실패 관련 정보를 담으라](https://codingwell.tistory.com/157?category=1013497)

eunveloper replied:
제가 에러메세지나 코드를 외부클래스에 상수로 선언하여 사용하는 버릇이 있는것 같네요 ㅠㅠ 늘 이렇게 진행했던거 같아요.
다만 '범용성' 이라는 단어가 제일 중요한 포인트가 될거같아요.

지금같은 상황에서는 범용성이 없는 에러 메세지를 선언하니 오히려 결합도가 생기게 되는게 맞는거 같아요!
적재적소에 사용하는게 제일 어렵네요 😂😂😂

두번째 코멘트는 반영해볼게요! 저는 요구사항에 있는 에러메세지를 변경해서는 안된다고 생각했는데,
제가 마음껏 커스텀 해도 되는 부분이라면 첨부해주신 사이트 내용처럼 명확하게 에러 상황을 메세지로 남기는게 좋다고 생각해요 👍🏻

devyonghee commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -36,17 +39,30 @@ public Answer(User writer, Question question, String contents) {
-        this.writer = writer;
-        this.question = question;
-        this.contents = contents;
+    public void isSameOwner(User loginUser) throws CannotDeleteException {
```
다른 부분들에서는 잘해주셔서 실수하신 것 같지만 
메서드의 네이밍과 내부에서 처리하고 있는 행동이 매칭이 안된다고 생각됩니다. 🤔
`boolean` 을  반환하는 것 같지만 검증을 수행하고 있는데 혹시 네이밍을 수정해보시는 방향은 어떠신가요??

eunveloper replied:
앗 그렇네요! 🤣

devyonghee commented on src/main/java/qna/domain/Question.java:
```diff
@@ -39,12 +50,53 @@ public Question writeBy(User writer) {
         return this;
     }
 
+    public DeleteHistories delete(User loginUser) throws CannotDeleteException {
```
데이터를 제거하면서 checked 예외를 사용하고 있는 것 같아요 👍
그런데 저는 개인적으로 기본적인 프로그래밍 오류는 unchekced 예외로 사용하고 있는데요.  
기존에 존재하던 예외를 그대로 사용하셨고 이미 알고 계실수도 있겠지만 참고해보시면 좋을 것 같아 자료 남겨드립니다.
[복구할 수 있는 상황에는 검사 예외를, 프로그래밍 오류에는 런타임 예외를 사용하라](https://madplay.github.io/post/effectivejava-chapter10-exceptions#%EC%95%84%EC%9D%B4%ED%85%9C-70-%EB%B3%B5%EA%B5%AC%ED%95%A0-%EC%88%98-%EC%9E%88%EB%8A%94-%EC%83%81%ED%99%A9%EC%97%90%EB%8A%94-%EA%B2%80%EC%82%AC-%EC%98%88%EC%99%B8%EB%A5%BC-%ED%94%84%EB%A1%9C%EA%B7%B8%EB%9E%98%EB%B0%8D-%EC%98%A4%EB%A5%98%EC%97%90%EB%8A%94-%EB%9F%B0%ED%83%80%EC%9E%84-%EC%98%88%EC%99%B8%EB%A5%BC-%EC%82%AC%EC%9A%A9%ED%95%98%EB%9D%BC)

eunveloper replied:
아! 동일한 의견입니다. checked 예외를 사용하면 try / catch 나 throws 처리를 해줘야하는데,
사실 지금같은 케이스들은 그런 필요가 없이 바로 unchecked 예외로 적절한 에러메세지를 리턴하는게 올바른 경우니까요! 👍🏻

devyonghee commented on src/main/java/qna/domain/Question.java:
```diff
@@ -39,12 +50,53 @@ public Question writeBy(User writer) {
+    }
+
+    public boolean isDeletedStatusAnswers() {
+        for (Answer answer : answers) {
+            if (!answer.isDeleted()) {
+                return false;
+            }
+        }
+        return true;
```
`indent` 가 두단계로 되어 있어서 조금 아쉬운 부분 같습니다. 😅
다음과 같이 stream 을 활용해보시면 명확하고 간결하게 변경할 수 있을 것 같은데 어떠신가요?
알고 계실 수도 있지만 참고하시면 좋을 것 같아 이외에도 스트림의 다양한 기능들에 대한 자료 남겨드리도록 하겠습니다!
[스트림 활용](https://velog.io/@kanamycine/%EB%AA%A8%EB%8D%98-%EC%9E%90%EB%B0%94%EC%9D%B8-%EC%95%A1%EC%85%98-chp-5.-%EC%8A%A4%ED%8A%B8%EB%A6%BC-%ED%99%9C%EC%9A%A9-1#03-%EA%B2%80%EC%83%89%EA%B3%BC-%EB%A7%A4%EC%B9%AD)
```suggestion
        return answers.stream()
                .allMatch(Answer::isDeleted);
```

eunveloper replied:
아닛! 전체적으로 보면서 indent 두단계인걸 확인했는데 이렇게 반영이 안된부분이 있다니 ㅠ_ㅠ
좀더 꼼꼼히 살펴봐야겠어요!

devyonghee commented on src/main/java/qna/domain/Question.java:
```diff
@@ -39,12 +50,53 @@ public Question writeBy(User writer) {
+    private DeleteHistories generateDeleteHistories() {
+        DeleteHistories deleteHistories = new DeleteHistories();
+        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, id, writer, LocalDateTime.now()));
+        for (Answer answer : answers) {
+            answer.setDeleted(true);
+            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
+        }
```
답변의 상태를 변경하고  답변 기록들을 생성하는 곳이 `Question` 클래스의 역할이 올바른 방향일까요??🤔🤔🤔
답변에서 수행되어야 하는 역할이 더 적합하다고 생각하는데 혹시 이 부분에 대해서 어떻게 생각하시나요??

eunveloper replied:
맞는 이야기 인거 같아요!
객체에게 '메세지를 전달' 하여 객체가 직접 상태를 변경할 수 있는게 올바르겠네요!
그러면 Question 뿐만 아니라 다른 Answer Entity와 매핑관계에 있는 다른 Entity에서도 사용할 수 있기도 하고요!😋

devyonghee commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -38,24 +38,7 @@ public Question findQuestionById(Long id) {
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
-        deleteHistoryService.saveAll(deleteHistories);
+        DeleteHistories deleteHistories = question.delete(loginUser);
+        deleteHistoryService.saveAll(deleteHistories.getList());
```
서비스 레이어가 많이 얇아진 것 같아 좋은 것 같아요 👍👍👍

devyonghee commented on src/main/java/qna/domain/DeleteHistories.java:
```diff
@@ -0,0 +1,17 @@
+
+    public List<DeleteHistory> getList() {
+        return deleteHistories;
+    }
```
일급 컬렉션을 적용해주셨네요 👍👍

devyonghee commented on src/main/java/qna/domain/Question.java:
```diff
@@ -19,6 +27,9 @@ public class Question extends BaseTime {
     @JoinColumn(name = "writerId", foreignKey = @ForeignKey(name = "fk_question_writer"))
     private User writer;
 
+    @OneToMany(fetch = FetchType.EAGER, mappedBy = "question")
+    private List<Answer> answers = new ArrayList<>();
```
답변들 리스트에도 접근하기 위해 반복문이 여러번 사용되는 것 같은데요 🤔
이에 대한 응집도를 위해 답변들에도 일급 컬렉션을 적용해보는 방향은 어떻게 생각하시나요?🤔🤔🤔

eunveloper replied:
좋은 생각이신거 같아요!

devyonghee commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -60,21 +46,27 @@ void save() {
         assertAll(
                 () -> assertThat(answerRepository.findById(actual.getId()))
-                        .isPresent().get().extracting(Answer::getContents).isEqualTo(answer.getContents()),
+                        .isPresent()
+                        .get()
+                        .extracting(Answer::getContents)
+                        .isEqualTo(answer.getContents()),
```
피드백을 수용해주시고 수정해주셔서 감사합니다 🙇‍♂️🙏

devyonghee commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -60,21 +46,27 @@ void save() {
     @DisplayName("Answer 객체를 조회하면 데이터 여부에 따라 Optional 존재 여부가 다르다." +
             "또한 동일한 객체면 담긴 값도 동일하다.")
     void findByWriterId() {
+        Answer answer = generateAnswer();
```
테스트 픽스처에 대해 피드백을 수용해주셔서 감사합니다 🙇‍♂️🙏

devyonghee commented on src/test/java/qna/unit/QuestionDeleteUnitTest.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    @Test
+    void 질문데이터_삭제는_데이터의_상태를_변경() throws Exception {
```
테스트 이름을 `@DisplayName` 으로 지정해주시다가 메서드 명을 한글로 정의해주셨는데 
혼란을 없애기 위해 한 가지의 컨벤션을 유지해주시는 방향은 어떻게 생각하시나요??🤔

eunveloper replied:
아! 제가 다른과제랑 병렬작업을 하는데 이것저것 시도하면서 헷갈렸나봐요!
적어도 한 프로젝트에서는 컨벤션을 유지하는게 좋은 방향이라고 생각해요! 😋

devyonghee commented on src/test/java/qna/unit/QuestionDeleteUnitTest.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Test
+    void 질문데이터_삭제는_데이터의_상태를_변경() throws Exception {
+        assertThat(question.isDeleted()).isFalse();
```
이 검증은 무엇을 위한 검증인지 고민해보시면 좋을 것 같아요 🤔
개인적으로 질문이 생성된 결과에 대해 상태 검증을 수행하는 것이 올바르지 않을까 생각되는데 혹시 어떻게 생각하시나요?

eunveloper replied:
음 이 클래스 자체가 질문 삭제에 대한 작은 기능 요구사항들의 테스트 클래스였는데,

요구사항에 
`질문 데이터를 완전히 삭제하는 것이 아니라 데이터의 상태를 삭제 상태(deleted - boolean type)로 변경한다.`
가 있어서 이 기능을 테스트하고자 했어요!

devyonghee commented on src/test/java/qna/unit/QuestionDeleteUnitTest.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Test
+    void 질문과답변_삭제이력은_내역을_관리() throws Exception {
+        assertThat(question.delete(UserTest.JAVAJIGI).getList().size()).isGreaterThan(0);
```
알고 계실 수 있지만 다음과 같이 검증하시는 것을 의도하신 것 같아요 😆
```suggestion
        assertThat(question.delete(UserTest.JAVAJIGI).getList()).isNotEmpty();
```

eunveloper replied:
오호 그러네요!
저 메소드를 확인해보니 null 체크와 Collection 형태면 IsEmpty() 까지 모두 확인해주는군요! 👍🏻

devyonghee commented on src/test/java/qna/unit/QuestionDeleteUnitTest.java:
```diff
@@ -0,0 +1,63 @@
+        question.addAnswer(new Answer(1L, UserTest.SANJIGI, question, "Answers Contents1"));
+
+        assertThatThrownBy(() -> question.delete(UserTest.JAVAJIGI))
+                .isInstanceOf(CannotDeleteException.class);
```
테스트 픽스처가 잘못 설정되어 답변을 우선 삭제해야 하는 상황이 발생된 것 같은데요 🤔
공통으로 사용되는 테스트 픽스처가 다른 테스트에도 영향을 받지 않도록 
테스트 구성을 고민해보시면 좋을 것 같은데 어떻게 생각하시나요?🤔🤔

devyonghee commented on src/test/java/qna/unit/QuestionDeleteUnitTest.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Test
+    void 로그인사용자_질문사용자_동일해야_삭제가능() throws Exception {
+        assertThat(question.delete(UserTest.JAVAJIGI)).isNotNull();
+        assertThatThrownBy(() -> question.delete(UserTest.SANJIGI))
+                .isInstanceOf(CannotDeleteException.class);
```
이미 삭제된 질문에 대해 다시 삭제를 수행하는데 
로그인 사용자가 잘못되었다는 동일한 메세지가 반환될 것 같은데요 🤔
잘못된 상태에 대한 요청 예외가 반환되어야 할 것 같은데 혹시 이부분에 대해 어떻게 생각하시나요?🤔🤔

devyonghee reviewed (changes requested):
안녕하세요 김은혜님!!
역시 3단계 질문 삭제하기 리팩터링도 잘 구현해주신 것 같아요!!

그런데 도메인에 불필요한 `setter`, `getter` 메서드가 남아있지 않은지, 
서비스에서 호출하면 안되는 메서드들이 노출되어 있지 않은지 고민해보시면 좋을 것 같아요 🤔 
도메인 클래스에 대해 전체적인 부분을 다시 한번 살펴봐주시면 감사하겠습니다 🙇‍♂️🙇‍♂️🙇‍♂️

이외에도 추가적으로 고민해보시면 좋을 것 같은 부분들에 대해 코멘트들을 남겨드렸는데요.
마지막 단계이다 보니 우선 request change 로 남겨드리겠습니다.
이에 대한 의견 남겨주시면 감사하겠습니다 🙇‍♂️🙇‍♂️
혹시 궁금하시거나 이상한 점이 보이신다면 편하게 DM 이나 코멘트 남겨주시기 바랍니다. 😆

> Entity를 생성할때 생성자로 넘어오는 문자열 데이터들도 포장하여 사용하는것이 좋을까요?

db 쿼리가 발생되기 아전에 필요한 조건(nullable, length, 빈 문자열, 특수문자 등)을 검증하는 부분, 컬럼에 대한 타입 변경 등 다양한 역할을 캡슐화 할 수 있기 때문에 
개인적으로  문자열 데이터도 포장하면 좋지 않을까 라는 생각을 하고 있습니다. 🤔🤔
다만, 모든 데이터 타입을 래퍼 클래스로 감싸게 되면 비용이 많이 들 수 있기 때문에 적절한 트레이드 오프가 필요할 것 같습니다. 😆

eunveloper committed:
refactor: 코드 리팩토링

- 프로그래밍 요구사항 맞추기
- 기타 추가 리팩토링이 필요한 부분 리팩토링하기

eunveloper committed:
refactor: 예외처리 관련 수정

- 불필요 Exception 제거
- 예외 상태메세지 명시
- 에러메세지 (외부 클래스 상수 결합 제거)

eunveloper committed:
refactor: 메서드명 수정

- is(boolean) -> checked(void)

eunveloper committed:
refactor: exception 적절하게 구성

- checked -> unchecked 로 변경 (프로그래밍 오류시)
- 불필요 throws 문 제거
- 사용 exception 변경에 따른 테스트코드 변경

eunveloper committed:
refactor: indent 축소

- indent 2단계인곳 stream을 활용하여 축소

eunveloper committed:
refactor: answer 일급콜렉션 생성

- Answers 객체 생성
- 객체에게 정확하게 행위를 위임하게 메소드 위치 변경
- 메소드명 리팩토링
- 테스트코드 변경

eunveloper committed:
refactor: 미사용 메소드 제거

eunveloper committed:
refactor: 테스트코드 리팩토링

- 테스트메소드 컨벤션 통일화
- 리스트 검증방식 변경

devyonghee commented on src/main/java/qna/UnAuthorizedException.java:
```diff
@@ -1,26 +1,19 @@
-                                 boolean writableStackTrace) {
-        super(message, cause, enableSuppression, writableStackTrace);
-    }
+    public UnAuthorizedException(User writer, User requester) {
+        super(String.format(
+                "질문을 삭제할 권한이 없습니다. [질문작성자: %s] [삭제요청자: %s]",
+                writer.getUserId(), requester.getUserId()));
```
내부에서 사용되는 메세지와 예외의 타입이 매칭이 어색한 것 같은데 혹시 이 부분에 대해서 어떻게 생각하시나요??
외부에서 어떤 경우에 사용될 지 미리 예상된 예외로 정의가 된 것 같아서요 🤔🤔🤔

devyonghee commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -57,12 +71,12 @@ public void toQuestion(Question question) {
 
-    public Long getId() {
-        return id;
+    public void changeDeleted() {
+        this.deleted = true;
```
이 메소드의 접근 제어자가 `public` 로 지정되어 있는데요. 
외부에서 검증 로직을 수행하지 않고 상태만 변경할 수 있을 것 같은데요. 
혹시 내부에서만 사용되고 호출되어야 하는 메서드라면 다른 접근 제어자로 변경해보시는 방향은 어떻게 생각하시나요??

devyonghee commented on src/main/java/qna/domain/Question.java:
```diff
@@ -39,38 +47,49 @@ public Question writeBy(User writer) {
+    public DeleteHistories delete(User loginUser) {
+        this.validationDeleteQuestionRequestUser(loginUser);
+        answers.validationDeleteAnswersRequestUser(loginUser);
+
+        setDeleted(true);
+
```
이 메서드는 내부에서만 호출되어야 할 것 같은데
`setDeleted` 메서드의 접근제어자가 어떤 게 적절한지 고민해보시면 좋을 것 같아요!!🤔🤔

devyonghee commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,41 @@
+        }
+    }
+
+    public List<DeleteHistory> generateDeleteHistories() {
+        List<DeleteHistory> deleteHistories = new ArrayList<>();
+        for (Answer answer : answers) {
+            answer.changeDeleted();
+            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
+        }
+        return deleteHistories;
+    }
```
메서드 명이 삭제 기록만 생성할 것 같은데 내부에서 상태도 같이 변경하고 있는데요.🤔🤔
메서드명과 내부에서 하고 있는 동작을 고민해보시면 좋을 것 같아요😆

devyonghee commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -36,17 +37,30 @@ public Answer(User writer, Question question, String contents) {
     public Answer(Long id, User writer, Question question, String contents) {
         this.id = id;
 
+        validationAuthorized(writer);
+        validationNotFounded(question);
```
반환되는 값도 없고 내부에서 검증만을 수행하고 있는 것 같은데요 🤔
개인적으로 저는 엘레강트 오브젝트 책에서 소개하고 있는 메서드명 컨벤션을 이용하고 있는데
메서드명이 명사라서 조금 어색하다고 느끼는데 혹시 어떻게 생각하시나요??🤔

알고 계실 수 도 있지만 제가 참고 해본 자료 첨부 해드리도록 하겠습니다.
[메서드 이름을 신중하게 선택하세요](https://jessyt.tistory.com/141)

devyonghee reviewed (approved):
안녕하세요 은혜님!!
마지막 단계까지 피드백 반영해주시느라 고생 많으셨습니다 🙇‍♂️🙇‍♂️🙇‍♂️
질문 답변 삭제 상태 대해 고민해주신 흔적들이 많이 보이는 것 같아요 👍👍

은혜님께서 코드를 잘 작성해주셔서 
제가 리뷰어로써 많은 피드백을 드리지 못한 것 같네요 🥲

추가적으로 한번 더 고민해보시면 좋을 것 같은 부분들 코멘트 남겨드렸습니다!
JPA 미션 진행하시느라 정말 고생 많으셨고
남은 미션도 열심히 하셔서 꼭 좋은 결과 있으셨으면 좋겠습니다 😆👍
